src/charm.py

```diff
@@ -18,3 +24,56 @@ def __init__(self, *args):
             args: Arguments passed to the CharmBase parent constructor.
         """
         super().__init__(*args)
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.saml_relation_created, self._on_relation_created)
+
+    def _on_config_changed(self, _) -> None:
+        """Handle changes in configuration."""
+        self.unit.status = ops.MaintenanceStatus("Configuring charm")
+        try:
+            self._charm_state = CharmState.from_charm(charm=self)
+            self._saml_integrator = SamlIntegrator(charm_state=self._charm_state)
+        except CharmConfigInvalidError as exc:
+            self.model.unit.status = ops.BlockedStatus(exc.msg)
+            return
+        if self.model.unit.is_leader():
+            for relation in self.model.relations["saml"]:
+                relation.data[self.model.app].update(self.dump_saml_data())
```

Comment:
Should the saml data come from the state?

Reply:
Are you referring to `dump_saml_data` being defined in the `SamlIntegrator` class? `dump_saml_data` takes that data and formats it according to the relation expectation, which is charm logic

Reply:
So relation data generally should be abstracted via the state and be made available via the state so that modules that need it can just take the state as an argument

Reply:
The state is the source of the data that is exposed in this relation. I'm not sure I'm getting your point.

Reply:
In my opinion as stated in other comments, this should be managed in the provider library

Reply:
Let's discuss this once the PR with the libary lands